incubator/hnc/test/conversion/conversion_test.go

```diff
@@ -0,0 +1,144 @@
+package conversion
+
+import (
+	"os"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+
+	. "sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils"
+)
+
+const namspacePrefix = "e2e-conversion-test-"
+
+func TestConversion(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	SetDefaultEventuallyTimeout(time.Second * 2)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"HNC v1alpha2 Conversion Test Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = Describe("Conversion from v1alpha1 to v1alpha2", func() {
+	const (
+		// hncVersion has to be a version using v1alpha1.
+		hncVersion = "v0.5.0"
+		nsA    = namspacePrefix + "a"
+		nsB    = namspacePrefix + "b"
+		nsC    = namspacePrefix + "c"
+	)
+
+	var (
+		crds = []string{"subnamespaceanchors.hnc.x-k8s.io","hierarchyconfigurations.hnc.x-k8s.io", "hncconfigurations.hnc.x-k8s.io"}
+	)
+
+	BeforeEach(func() {
+		// We need GOPATH here to install kubectl-hns plugin.
+		CheckGoPath()
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Tear down HNC for both the specified version and at HEAD.
+		TearDownHNC(hncVersion)
+		// Wait for HNC to be fully torn down.
+		ensureHNCCleared(crds)
+	})
+
+	AfterEach(func() {
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Only tear down HNC at HEAD, since that's what we just deployed.
+		TearDownHNC("")
+	})
+
+	It("should convert CRD versions properly and HNC should still work", func() {
+		// Set up v1alpha1 and verify everything is fine before conversion.
+		GinkgoT().Log("Set up v1alpha1")
+		setupV1alpha1(hncVersion)
+		// Wait for the validating webhooks to be ready.
+		ensureVWHReady()
```

Comment:
Maybe combine these two? `setupV1alpha1` shouldn't return until it's ready to use.

You could also argue that we should _delete_ the VWHes. They're not needed since these are not e2e tests, and deleting them makes it easier/possible to create objects with conditions.

Reply:
Also, everything up to this line should probably be part of `BeforeEach`, right? They'll be the same for every test in this suite.

Reply:
Updated.